lib/aux_js.cpp

```diff
@@ -222,23 +223,76 @@ bool startVisualization(const std::string input, const std::string data_type,
    }
 
    SendExposeEvent();
-   return true;
+   return 0;
 }
 
-
-int updateVisualization(std::string data_type, std::string stream)
+//
+// StreamState::ReadStream requires a list of unique_ptr to istream and since
+// we cannot directly pass a list of string we need to repack the strings into
+// a new list.
+//
+// each string in streams must start with `parallel <nproc> <rank>'
+//
+using StringArray = std::vector<std::string>;
+int processParallelStreams(StreamState & state, const StringArray & streams,
+                           std::stringstream * commands = nullptr)
 {
-   std::stringstream ss(stream);
+   //std::cerr << "got " << streams.size() << " streams" << std::endl;
+   // HACK: match unique_ptr<istream> interface for ReadStreams:
+   std::vector<std::stringstream> sstreams(streams.size());
+   StreamCollection istreams(streams.size());
+   for (int i = 0; i < streams.size(); ++i)
+   {
+      sstreams[i] = std::stringstream(streams[i]);
+      // pull off the first list
+      std::string word;
+      int nproc, rank;
+      sstreams[i] >> word >> nproc >> rank;
+      //std::cerr << "packing " << rank+1 << "/" << nproc << std::endl;
+      istreams[i] = std::unique_ptr<std::istream>(&sstreams[i]);
+   }
 
-   if (data_type != "solution")
+   const int field_type = state.ReadStreams(istreams);
+
+   if (commands)
    {
-      std::cerr << "unsupported data type '" << data_type << "' for stream update" <<
-                std::endl;
-      return 1;
+      commands->seekg(istreams[0]->tellg());
    }
 
-   StreamState new_state;
-   new_state.ReadStream(ss, data_type);
+   // HACK: don't let unique_ptr free the data
+   for (int i = 0; i < streams.size(); ++i)
+   {
+      istreams[i].release();
```

Comment:
Are these streams being used elsewhere outside this function?

Reply:
No they are not.

Ideally we can come up with a way to get a `std::vector<std::unique_ptr<std::istream>>` view of the `std::vector<std::string>` argument but I'm not sure how to do that. Any ideas? The other option is to change the signature of `ReadStream` but that would mean changing `glvis.cpp`

Reply:
I guess I was asking more about the use of `unique_ptr::release()`; I think the use of `std::stringstream` is fine given that we also have to handle the case of live sockets via mfem's `socketstream` when running as a desktop application. It also looks like the `mfem::Mesh` and `GridFunction` ctors expect an `std::istream`-like argument.

Reply:
Ah! Good question. The `unique_ptr`s are referencing the stack allocated `streamstream`s and I don't want `unique_ptr` to try to destruct them since that would be a double-free. I'm not sure if we can improve, thoughts ?